### PR TITLE
Tutorial for MT marathon

### PR DESCRIPTION
**Pull request by tdomhan**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,16 +18,22 @@ Sockeye requires:
 
 ## Installation
 
-### For AWS DeepLearning AMI users
+There are several options for installing Sockeye and it's dependencies. Below we list several alterantives and the
```
Comment (fhieber) [inline README.md]: typo: alternatives

File: README.md [hunk 2]
```diff
@@ -18,16 +18,22 @@ Sockeye requires:
...
+corresponding instructions.
 
-AWS DeepLearning AMI users only need to run the following line to install sockeye:
+### Either: AWS DeepLearning AMI
```
Comment (fhieber) [inline README.md]: I would remove the "Either" and "Or" below

File: README.md [hunk 3]
```diff
@@ -18,16 +18,22 @@ Sockeye requires:
...
+[AWS DeepLearning AMI](https://aws.amazon.com/amazon-ai/amis/) users only need to run the following line to install sockeye:
 
 ```bash
 > sudo pip3 install sockeye --no-deps
 ```
 
-For other environments, you can choose between installing via pip or directly from source.
-### pip
+For other environments, you can choose between installing via pip or directly from source. Note that for the
+remaining instructions to work you will need to used `python3` instead of `python` and `pip3` instead of `pip`.
```
Comment (fhieber) [inline README.md]: used -> use

File: README.md [hunk 4]
```diff
@@ -37,8 +43,8 @@ For other environments, you can choose between installing via pip or directly fr
 
 #### GPU
 
-If you want to run sockeye on a GPU you need to make sure your version of Apache MXNet contains the GPU code. Depending on your
-version of CUDA you can do this by running the following for CUDA 8.0:
+If you want to run sockeye on a GPU you need to make sure your version of Apache MXNet Incubating contains the GPU code.
```
Comment (fhieber) [inline README.md]: GPU bindings maybe?

File: README.md [hunk 5]
```diff
@@ -52,7 +58,7 @@ or the following for CUDA 7.5:
 > rm requirements.gpu-cu75.txt
 ```
 
-### From Source
+### Or: From Source
```
Comment (fhieber) [inline README.md]: remove or

File: README.md [hunk 6]
```diff
@@ -64,8 +70,9 @@ after cloning the repository from git.
 
 #### GPU
 
-If you want to run sockeye on a GPU you need to make sure your version of Apache MXNet contains the GPU code. Depending on your
-version of CUDA you can do this by running the following for CUDA 8.0:
+If you want to run sockeye on a GPU you need to make sure your version of Apache MXNet
```
Comment (fhieber) [inline README.md]: if you change above, should be the same here

File: README.md [hunk 7]
```diff
@@ -136,3 +140,8 @@ write translations to STDOUT.
 
 For more detailed examples check out our user documentation.
 
+
+## Step by step tutorial
+
+More detailed step by step tutorials can be found in the
```
Comment (fhieber) [inline README.md]: step-by-step

File: sockeye/train.py [hunk 8]
```diff
@@ -81,10 +81,7 @@ def _dict_difference(dict1: Dict, dict2: Dict):
 
 def main():
     params = argparse.ArgumentParser(description='CLI to train sockeye sequence-to-sequence models.')
-    arguments.add_io_args(params)
-    arguments.add_model_parameters(params)
-    arguments.add_training_args(params)
-    arguments.add_device_args(params)
+    arguments.add_train_cli_args(params)
```
Comment (fhieber) [inline sockeye/train.py]: nice :)

File: test/unit/test_arguments.py [hunk 9]
```diff
@@ -168,8 +168,123 @@ def test_inference_args(test_params, expected_params):
     _test_args(test_params, expected_params, arguments.add_inference_args)
 
 
+# Make sure that the parameter names and default values used in the tutorials do not change without the tutorials
+# being updated accordingly.
+@pytest.mark.parametrize("test_params, expected_params, expected_params_present", [
```
Comment (fhieber) [inline test/unit/test_arguments.py]: nice example of test-driven development :)

File: tutorials/README.md [hunk 10]
```diff
@@ -0,0 +1,26 @@
+# Step by step tutorials
```
Comment (fhieber) [inline tutorials/README.md]: Step-by-Step?

File: tutorials/README.md [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+## Setup
+
+For running the examples in the tutorials you can either manually install Sockeye, it's dependencies and additional
+tools used or use the Dockerfile we provide, which comes with all batteries included.
+
+### Either: Dockerfile
```
Comment (fhieber) [inline tutorials/README.md]: You like these  :)

File: tutorials/README.md [hunk 12]
```diff
@@ -0,0 +1,26 @@
+# Step by step tutorials
+
+## Setup
+
+For running the examples in the tutorials you can either manually install Sockeye, it's dependencies and additional
```
Comment (fhieber) [inline tutorials/README.md]: "[...] either install Sockeye manually and its dependencies and additional tools, or use the Dockerfile provided which comes with everything included."

File: tutorials/README.md [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+tools used or use the Dockerfile we provide, which comes with all batteries included.
+
+### Either: Dockerfile
+
+Coming soon...
+
+<!--we provide a docker file that already includes all prerequisites for running the tutorials.
+If you're using the provided Dockerfile simply skip the `Setup` sections in the tutorials below.-->
```
Comment (fhieber) [inline tutorials/README.md]: I would avoid contractions in documentation

Comment (fhieber) [reply tutorials/README.md]: "sections" -> "section"?

File: tutorials/README.md [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+### Or: Manual installation
+If you do not use the provided Dockerfile you can follow the [installation instructions](../README.md#installation)
+for how to manually install Sockeye and all dependencies. The tutorials below might have additional dependencies that
```
Comment (fhieber) [inline tutorials/README.md]: would remove "for how"

File: tutorials/README.md [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+will mentioned at the beginning of each tutorial.
```
Comment (fhieber) [inline tutorials/README.md]: will be mentioned

File: tutorials/README.md [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+
+## Tutorials
+
+Below is the full list of tutorials we provide. We recommend going through them in order as they will gradually
+introduce the different concepts and parameters used for training and translation.
```
Comment (fhieber) [inline tutorials/README.md]: "introduce different concepts and parameters [...]"

File: tutorials/seqcopy/README.md [hunk 17]
```diff
@@ -0,0 +1,140 @@
+# Sequence copy model
+This tutorial will show you the basic usage of Sockeye on a task simpler than Machine Translation: copying a sequence.
```
Comment (fhieber) [inline tutorials/seqcopy/README.md]: on a simple task: copying a sequence.

File: tutorials/seqcopy/README.md [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+We will generate sequences consisting of digits of variable lengths.
+The task is then to train a model that copies the sequence from the source to the target.
+This task is on the one hand difficult enough to be interesting and on the other and allows for quickly training a model.
+
+## Setup
+For this tutorial we assume that you have successfully [installed](../../README.md#installation) Sockeye.
+Other than that there are no additional dependencies.
```
Comment (fhieber) [inline tutorials/seqcopy/README.md]: In that case this sentence is not needed

File: tutorials/seqcopy/README.md [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+We will be using scripts from the Sockeye repository, so you should either clone the repository or manually download 
+the scripts.
+Just as a reminder: Everything is run using Python 3, so depending on your setup you may have to replace `python` with
+`python3` below.
+All of the commands below assume you're running on a CPU.
```
Comment (fhieber) [inline tutorials/seqcopy/README.md]: remove contraction?

File: tutorials/seqcopy/README.md [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+As a first step we will generate a synthetic data set consisting of random sequences of digits.
+These sequences are then split into disjoint training and development sets.
+Run the following command to create the data set:
+
+```bash
+python genseqcopy.py
+```
+
+After running this script you have a training (`train.source`, `train.target`) and a development data set
+(`dev.source`, `dev.target`). The sequences that were generate look like this:
```
Comment (fhieber) [inline tutorials/seqcopy/README.md]: The generated sequences will look like this:

File: tutorials/seqcopy/README.md [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+                         --num-embed 32 \
+                         --rnn-num-hidden 64 \
+                         --attention-type dot \
+                         --use-cpu \
+                         --metrics perplexity accuracy \
+                         --max-num-checkpoint-not-improved 3 \
+                         -o seqcopy_model
+```
+
+This will train a 1-layer RNN model with a bidirectional LSTM as the encoder side and a uni-directional LSTM
```
Comment (fhieber) [inline tutorials/seqcopy/README.md]: "as the encoder" or "on the encoder side"

File: tutorials/seqcopy/README.md [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+[INFO:sockeye.training] Training started.
+[INFO:sockeye.callback] Early stopping by optimizing 'perplexity'
+[INFO:root] Epoch[0] Batch [50]  Speed: 683.23 samples/sec perplexity=14.104128 accuracy=0.092011
+[INFO:root] Epoch[0] Batch [100] Speed: 849.97 samples/sec perplexity=13.036482 accuracy=0.096760
+...
+```
+With a vocabulary of size 10 this essentially means that the model is guessing randomly.
+As training progresses we see that after around 14 epochs the accuracy goes up to ~1.0 and the perplexity down to ~1.0.
+Sockeye performs early stopping based on the validation metrics tracked when checkpointing.
+Once the validation metrics have not improve for several checkpoints the training is stopped.
```
Comment (fhieber) [inline tutorials/seqcopy/README.md]: improved

File: tutorials/seqcopy/README.md [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+How many validation non-improved checkpoints to tolerate can be set adjusted `--max-num-checkpoint-not-improved`.
```
Comment (fhieber) [inline tutorials/seqcopy/README.md]: The number of tolerated non-improving checkpoints can be adjusted ...

File: tutorials/seqcopy/README.md [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+  
+        7 6 7 7 10 2 0 8 0 5 7 3 5 6 4 0 0 2 10 0
+
+```
+
+Note that the model was trained on sequences consisting of between 10 and 30 characters.
+Therefore, the model will most likely have some difficulties with sequences shorter than 10 characters.
+By default Sockeye will read sentence from stdin and print the translations on stdout.
+
+Internally Sockeye will run a beam search in order to (approximately) find the translation  with the highest
```
Comment (fhieber) [inline tutorials/seqcopy/README.md]: superfluous whitespace

File: tutorials/wmt/README.md [hunk 25]
```diff
@@ -0,0 +1,251 @@
...
+[Conference on Machine Translation (WMT) 2017](http://www.statmt.org/wmt17/).
+
+## Setup
+
+Sockeye expects already tokenized data as the input.
+For this tutorial we use data that has already been tokenized for us.
+However, keep this in mind for any other data set you want to use with Sockeye.
+In addition to tokenization we will splits words into subwords using Byte Pair Encoding (BPE).
+In order to do so we use a tool called [subword-nmt](https://github.com/rsennrich/subword-nmt).
+Run the following commands to set the tool up:
```
Comment (fhieber) [inline tutorials/wmt/README.md]: to set up the tool

File: tutorials/wmt/README.md [hunk 26]
```diff
@@ -0,0 +1,251 @@
...
+                        --rnn-num-hidden 512 \
+                        --attention-type dot \
+                        --max-seq-len 60 \
+                        --monitor-bleu 500 \
+                        --use-tensorboard \
+                        --use-cpu \
+                        -o wmt_model
+```
+
+This will train a 1-layer bi-LSTM encoder, 1 layer LSTM decoder with dot attention.
```
Comment (fhieber) [inline tutorials/wmt/README.md]: 1-layer

File: tutorials/wmt/README.md [hunk 27]
```diff
@@ -0,0 +1,251 @@
...
+Unless you specify a different optimizer (`--optimizer`) [Adam](https://arxiv.org/abs/1412.6980) will be used.
+Additionally, you can control the batch size (`--batch-size`), the learning rate schedule (`--learning-rate-schedule`)
+and other parameters relevant for training.
+
+Training will run until the validation perplexity stops improving.
+Alternatively, you can track BLEU on the validation set (`--optimized-metric bleu`).
+Sockeye will then at every checkpoint start a decoder in a separate process running on the same device as training.
+To make sure the decoder finishes before the next checkpoint one can subsample the validation set for
+BLEU score calculation.
+For example `--monitor-bleu 500` will calculate BLEU on a random subset of 500 sentences.
```
Comment (fhieber) [inline tutorials/wmt/README.md]: maybe note that this random set is sampled once and based on a fixed random seed

File: tutorials/wmt/README.md [hunk 28]
```diff
@@ -0,0 +1,251 @@
...
+written to disk.
+
+## Translation
+
+When translating with Sockeye it is important to keep in mind that it expects the same types of input as seen during
+training. For this tutorial we fed in subword units that were obtained through a Byte Pair Encoding.
+Therefore, we need to apply the same type of preprocessing before feeding a sentence into Sockeye.
+All symbols that have not been seen during training will be replaced by an `<unk>` symbol.
+When the `<unk>` symbol was observed during training one can that the model will also produce this symbol on the output.
+Note though that bcause of the wy we do the preprocessing with BPE above, the model will not actually observe any
```
Comment (fhieber) [inline tutorials/wmt/README.md]: "way"

File: README.md (hunk 1 above)
Comment (tdomhan) [reply README.md]: fixed

File: README.md (hunk 2 above)
Comment (tdomhan) [reply README.md]: I just thought that it wasn't entirely clear there these were different options and I wanted to make this as clear as possible.

File: README.md (hunk 3 above)
Comment (tdomhan) [reply README.md]: fixed

File: README.md (hunk 4 above)
Comment (tdomhan) [reply README.md]: sure

File: README.md (hunk 7 above)
Comment (tdomhan) [reply README.md]: changed.

File: tutorials/README.md (hunk 10 above)
Comment (tdomhan) [reply tutorials/README.md]: fixed.

File: tutorials/README.md (hunk 13 above)
Comment (tdomhan) [reply tutorials/README.md]: changed

File: tutorials/README.md (hunk 14 above)
Comment (tdomhan) [reply tutorials/README.md]: sure

File: tutorials/README.md (hunk 15 above)
Comment (tdomhan) [reply tutorials/README.md]: changed.

File: tutorials/README.md (hunk 16 above)
Comment (tdomhan) [reply tutorials/README.md]: changed

File: tutorials/seqcopy/README.md (hunk 19 above)
Comment (tdomhan) [reply tutorials/seqcopy/README.md]: sure

File: tutorials/seqcopy/README.md (hunk 20 above)
Comment (tdomhan) [reply tutorials/seqcopy/README.md]: changed

File: tutorials/seqcopy/README.md (hunk 21 above)
Comment (tdomhan) [reply tutorials/seqcopy/README.md]: changed

File: tutorials/seqcopy/README.md (hunk 22 above)
Comment (tdomhan) [reply tutorials/seqcopy/README.md]: changed

File: tutorials/seqcopy/README.md (hunk 23 above)
Comment (tdomhan) [reply tutorials/seqcopy/README.md]: changed

File: tutorials/seqcopy/README.md (hunk 24 above)
Comment (tdomhan) [reply tutorials/seqcopy/README.md]: good catch!

File: tutorials/wmt/README.md (hunk 25 above)
Comment (tdomhan) [reply tutorials/wmt/README.md]: changed

File: tutorials/wmt/README.md (hunk 26 above)
Comment (tdomhan) [reply tutorials/wmt/README.md]: changed

File: tutorials/wmt/README.md (hunk 27 above)
Comment (tdomhan) [reply tutorials/wmt/README.md]: done

File: tutorials/wmt/README.md (hunk 28 above)
Comment (tdomhan) [reply tutorials/wmt/README.md]: fixed.
